oat-sa/extension-lti-test-review
# Feature/tr 3180/lti review panel fizzy variant

https://oat-sa.atlassian.net/browse/TR-3180

LTI Review Panel - implementation of new design [[Invision](https://invis.io/V411RRJ8TRJM#/462582285_Test_Review_Panel_Scroll)], based on previous [POC PR](https://github.com/oat-sa/extension-lti-test-review/pull/99). Best to read that PR first if you are not familiar.

### Depends on
- [x] https://github.com/oat-sa/tao-core-ui-fe/pull/409
- [ ] https://github.com/oat-sa/tao-core/pull/3315

### What has changed from the POC?
- not much, I'm still using the `1: yes, 2: yes, 3: yes` options
- `stepOverview` list component was moved to tao-core-ui (renamed as `itemButtonList`)
- minor styles fixes
- added unit tests for new panel
- fixed BE bug about informational status
- ~~fixed FE bug/feature where skipped items were shown with red "incorrect" icon~~ **will revert this part after BA/UX discussion**
  - ~~not 100% sure about this, [Slack discussion](https://oatsa.slack.com/archives/C02SBELRSEB/p1641375459000200)~~
  - ~~note: 2 existing unit tests were also updated to reflect this change...~~
- ~~read the config file of `taoQtiTest` to get its values of `reviewLayout` and `displaySectionTitles`~~
  -  ~~not sure if that's ok, or if this extension should store its own config~~
- extension stores its own config (for `reviewLayout` and `dsplaySectionTitles`). We will override the defaults via a customer extension where needed.

TODO:
- [x] unit tests for `fizzyPanel`
- [x] could improve/rationalise `panel-data` (unified format)
- [x] check if tabfocus support can be removed here and added in `tao-core-ui`

### Playground
Env: http://test-tr-3180-reviewpanel.playground.kitchen.it.taocloud.org:42081/
Using recent [mexcbt-package composer](https://github.com/oat-sa/mexcbt-package/blob/2021.12.1-alpha/composer.json) as base

![Screenshot 2022-01-07 at 15 19 24](https://user-images.githubusercontent.com/43652944/148559848-dd6ff998-0c45-40eb-aef2-66b2589bb6c7.png)

### To test
1. A test has already been published on the deployed env. You can launch it from the [LTI 1.1 app](http://ltiapps.net/test/tc.php).
  - url: `http://test-tr-3180-reviewpanel.playground.kitchen.it.taocloud.org:42081/ltiDeliveryProvider/DeliveryTool/launch/eyJkZWxpdmVyeSI6Imh0dHA6XC9cL3Rlc3QtdHItMzE4MC1yZXZpZXdwYW5lbC5wbGF5Z3JvdW5kLmtpdGNoZW4uaXQudGFvY2xvdWQub3JnXC90YW8ucmRmI2k2MWRmZGM2ODc5MzYyMTg1MTg0Mjk4YzdlZjEwNDI2NiJ9`
  - key: `abc`
  - secret: `d6e369591edc11bf9fb25e988000730a18559a62`

2. Complete the delivery however you want (note there is a 1-minute timer on the last section).

3. From the location bar, make sure to copy the last param value (after the `=`): `?deliveryExecution={{DELIVERY_EXECUTION_URL}}`

4. Perform a second launch into LTI review mode:
  - url: `http://test-tr-3180-reviewpanel.playground.kitchen.it.taocloud.org:42081/ltiTestReview/ReviewTool/launch?execution={{DELIVERY_EXECUTION_URL}}`
  - key: same
  - secret: same

### To install
To install this thing locally, the [composers from the playground](https://github.com/oat-sa/kitchen-playground/tree/test/tr-3180-reviewpanel/files/delivery) should work.

- in `config/taoQtiTest/testrunner.conf.php`:
  - `plugins.review.reviewLayout` to be set to `'fizzy'` (or `'default'`)
  - `plugins.review.displaySectionTitles` to be set to `true` or `false`
- in `config/ltiTestReview/DeliveryExecutionFinderService.conf.php`:
  - `show_score` to be set to `true` or `false` (governs panel header & footer, and icons)
